### PR TITLE
pydantic 2  - migration

### PR DESCRIPTION
**Pull request by bertiqwerty**
The goal of this MR is to upgrade to Pydantic 2. For the initial commit I run `bump-pydantic`  and changed some things manually. There are still some open issues, though.

### CODE REVIEW

File: bofire/data_models/strategies/predictives/botorch.py
Comment (bertiqwerty) [inline bofire/data_models/strategies/predictives/botorch.py:78]: @jduerholt, is this correct?

File: bofire/data_models/strategies/predictives/botorch.py
Comment (bertiqwerty) [inline bofire/data_models/strategies/predictives/botorch.py:105]: @jduerholt, here again.

File: bofire/data_models/strategies/predictives/botorch.py
Comment (jduerholt) [reply bofire/data_models/strategies/predictives/botorch.py:78]: Hmm, I know that it definitely works, because this is always used when one does not provide any surrogates specs. I think, that, at least in the former pydantic versions, if not provided it will be None.

File: bofire/data_models/strategies/predictives/botorch.py
Comment (jduerholt) [reply bofire/data_models/strategies/predictives/botorch.py:105]: I only can say that it has worked :D

Comment (bertiqwerty) [reply bofire/data_models/strategies/predictives/botorch.py:105]: My question was if the change is correct or if it changes what happens :D

**Comment by jduerholt** [comment]
Todo list:
- review all model validators, we should switch everywhere to mode "after", this is more failsafe
- add validators for inputs and outputs objects to check for unique keys
- intotroduce commone type for unique str tuples

**Comment by jduerholt** [comment]
There are some bugs in pydantic2:
- Order of AnyFeature affects the correct deserialization
- cannot deal with `update_hyperparams` as abstract method ...

Things to consider:
- Look at `order_id` for features
- make one place for custom types

**Comment by jduerholt** [comment]
@bertiqwerty: Only failing tests are now due to failure in the DOE module since on the 25th of December a new formulaic version was released and this breaks some of our tests.

**Comment by jduerholt** [comment]
Hi @bertiqwerty,

tests are now passing, and I think the PR is ready for review.

What do you think, who should review it? 

Issues that are mentioned above which are still open are:
- create a central place for our custom types like `bofire/data_models/types.py`
- on several places in the code we use lists that constrain strings which have to be unique (examples are the `features` attribute from constraints, the categories attribute from categorical features etc.). In my opinion we should switch them all to  tuples and create a custom type for this.

But it could be better to do this in a seperate PR. 

What do you think?

Best,

Johannes

**Comment by bertiqwerty** [comment]
Hi @jduerholt , 

Thanks a lot! The person for review would probably be Benny. I will also have a look. Having a central `types.py` makes sense to me. We should do this in a separate PR.

File: bofire/data_models/constraints/nchoosek.py
Comment (bertiqwerty) [inline bofire/data_models/constraints/nchoosek.py:79]: Does this `# type ignore` do anything?

File: bofire/data_models/constraints/nchoosek.py
Comment (bertiqwerty) [inline bofire/data_models/constraints/nchoosek.py:85]: See above

File: bofire/data_models/strategies/predictives/botorch.py
Comment (bertiqwerty) [inline bofire/data_models/strategies/predictives/botorch.py:86]: Remove uncommented code

File: bofire/data_models/strategies/predictives/botorch.py
Comment (bertiqwerty) [inline bofire/data_models/strategies/predictives/botorch.py:150]: When I understood this correctly, `# type: ignore` correctly, it operates on the line level. We could at some point activate `reportUnnecessaryTypeIgnoreComment` in Pyright's options.
See https://github.com/microsoft/pyright/blob/main/docs/comments.md#line-level-diagnostic-suppression

File: bofire/data_models/surrogates/botorch_surrogates.py
Comment (bertiqwerty) [inline bofire/data_models/surrogates/botorch_surrogates.py:90]: Make an issue out of the TODO?

File: bofire/data_models/surrogates/fully_bayesian.py
Comment (bertiqwerty) [inline bofire/data_models/surrogates/fully_bayesian.py:13]: Maybe we don't need these `#type ignore`s due to `Annotated`?

File: bofire/surrogates/diagnostics.py
Comment (bertiqwerty) [inline bofire/surrogates/diagnostics.py:489]: Does `TODO[pydantic]` have a special meaning? And who is the one to replace stuff manually?

**Review by bertiqwerty — COMMENTED** [review]
Thanks Johannes. I cannot approve the PR because Github says authors cannot approve their own PRs. You can probably approve and merge it?

**Comment by jduerholt** [comment]
Hi @bertiqwerty, 

thx! Yes, I will go through it and then approve and merge. For me it works.

The idea with the `reportUnnecessaryTypeIgnoreComment` is a good one! Maybe we should also switch to a more recent pyright? We are still on version `1.1.305`.

Before than merging it in, should we make a last pydantic v1 release?

Best,

Johannes

**Comment by bertiqwerty** [comment]
Hi Johannes, 

I think we don't need another release with Pydantic 1 but I also don't have an objection. So I am fine either way.

Updating Pyright is a good idea. This often leads to some issues. So we could create a MR for this. I can look into this.

**Comment by jduerholt** [comment]
Ok, then we do not create a Pydantic 1 release. Would be nice if you can take care for the pyright update. But we should do it after merging in this PR. My plan is to have it in the latest tomorrow. Note that I moved the pyright config some time ago into the `pyproject.toml`, so the new flag for showing the unnecessary `ignore` statements should also go there, I think:

https://github.com/experimental-design/bofire/blob/33a2053e80c371791cb27364c442b83f866b1c09/pyproject.toml#L34

File: bofire/data_models/constraints/nchoosek.py
Comment (jduerholt) [reply bofire/data_models/constraints/nchoosek.py:79]: Good catch!

File: bofire/data_models/constraints/nchoosek.py
Comment (jduerholt) [reply bofire/data_models/constraints/nchoosek.py:85]: Good catch.

File: bofire/data_models/surrogates/botorch_surrogates.py
Comment (jduerholt) [reply bofire/data_models/surrogates/botorch_surrogates.py:90]: At some point, currently I am too lazy :D

File: bofire/surrogates/diagnostics.py
Comment (jduerholt) [reply bofire/surrogates/diagnostics.py]: Very good catch, i fixed it.

**Review by jduerholt — APPROVED** [review]

**Comment by jduerholt** [comment]
@bertiqwerty: it is now merged, you can now have a look at the pyright update ;)

**Comment by bertiqwerty** [comment]
I am in favor of making a new release with the new Pydantic version.

**Comment by bertiqwerty** [comment]
I can.

**Pull request merged**
